src/mono/mono/metadata/metadata.c

```diff
@@ -3434,7 +3434,8 @@ mono_metadata_get_canonical_generic_inst (MonoGenericInst *candidate)
 	MonoMemoryManager *mm = mono_mem_manager_get_generic (data.images, data.nimages);
 	collect_data_free (&data);
 
-	mono_mem_manager_lock (mm);
+	// Hashtable key equal func can take loader lock
```

Comment:
In `loader-internals.h` the following other caches do a deep comparison that might also have a problem:

* `gmethod_cache`
* `gsignature_cache`
* `gclass_cache`

There's also a ton of hashtables in `MonoWrapperCaches`, but I didn't check them - from my recollection they mostly do pointer comparison.

Reply:
I checked for other uses of `g_hash_table_new_full` with a comparison func that looks like it's doing deep type equality.  The only other places (besides the ones in the previous comment) are the 3 caches in `mini-generic-sharing.c` that call `mono_metadata_signature_equal` under the `gshared_lock()`.  I don't understand generic sharing well enough to tell if this needs to change to the loader lock, too. /cc @vargaz

Reply:
Thanks for taking a look. Not sure why you mention `gmethod_cache` and `gsignature_cache`, the equality function only does ref compares. Added a new commit to handle `gclass_cache`.  I don't think any of those wrapper creation methods are called while holding the loader lock. If that's the case we can't deadlock there.

Reply:
You're right. For some reason I thought `mono_metadata_generic_context_equal` does a deep equality - I think I probably wasn't careful when looking around the code.